File: services/api/app/models/election.py
"""Elections, versioned election rule sets, and election timelines
(sections 3, 6-60 baseline)."""

from __future__ import annotations

import uuid
from datetime import date, datetime

from sqlalchemy import Boolean, Date, DateTime, Enum, Float, ForeignKey, Integer, String, Text
from .types import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base, TimestampMixin, UUIDPrimaryKeyMixin
from .enums import ForecastRunStatus


class Election(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "elections"

    name_en: Mapped[str] = mapped_column(String(300))
    name_ar: Mapped[str] = mapped_column(String(300))
    election_type: Mapped[str] = mapped_column(String(50), default="legislative")
    scheduled_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    is_current: Mapped[bool] = mapped_column(Boolean, default=False)
    status: Mapped[str] = mapped_column(String(50), default="scheduled")

    rule_sets: Mapped[list["ElectionRuleSetORM"]] = relationship(back_populates="election")
    timeline_events: Mapped[list["ElectionTimelineEvent"]] = relationship(back_populates="election")


class ElectionRuleSetORM(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Persisted mirror of packages/election_rules_py.ElectionRuleSet.
    See section 3 — versioned so law changes never require a code change.
    """

    __tablename__ = "election_rule_sets"

    election_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("elections.id"))
    version: Mapped[str] = mapped_column(String(50))
    effective_from: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    effective_until: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    electoral_system: Mapped[str] = mapped_column(String(300))
    district_structure: Mapped[str] = mapped_column(String(300))
    total_seats: Mapped[int] = mapped_column(Integer)
    threshold_fraction: Mapped[float] = mapped_column(Float)
    allocation_method: Mapped[str] = mapped_column(String(50))

    reserved_seats: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    gender_quota: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    minimum_candidate_age: Mapped[int] = mapped_column(Integer)
    list_minimum_candidates: Mapped[int | None] = mapped_column(Integer, nullable=True)
    list_maximum_candidates: Mapped[int | None] = mapped_column(Integer, nullable=True)
    allows_individual_candidate_votes: Mapped[bool] = mapped_column(Boolean, default=False)

    source_document: Mapped[str] = mapped_column(Text)
    verified_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))

    election: Mapped["Election"] = relationship(back_populates="rule_sets")

    @property
    def majority_threshold(self) -> int:
        return (self.total_seats // 2) + 1


class ElectionTimelineEvent(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Registration / nomination / campaign / voting / results milestones
    (section 30 Election Calendar). Dates come from official CEC info, not
    hard-coded — see the seed script for the currently-cited baseline."""

    __tablename__ = "election_timelines"

    election_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("elections.id"))
    milestone: Mapped[str] = mapped_column(String(100))  # registration_open, nomination_close, ...
    label_en: Mapped[str] = mapped_column(String(300))
    label_ar: Mapped[str] = mapped_column(String(300))
    starts_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    ends_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    source_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sources.id"), nullable=True
    )

    election: Mapped["Election"] = relationship(back_populates="timeline_events")

File: apps/flutter_client/lib/models/election.dart
/// Election + versioned rule-set DTOs mirroring services/api/app/schemas/election.py.
/// See docs/ELECTORAL_RULES.md for why the rule set is versioned rather
/// than hard-coded.
class Election {
  final String id;
  final String nameAr;
  final String nameEn;
  final String electionType;
  final DateTime? scheduledDate;
  final bool isCurrent;
  final String status;

  const Election({
    required this.id,
    required this.nameAr,
    required this.nameEn,
    required this.electionType,
    required this.scheduledDate,
    required this.isCurrent,
    required this.status,
  });

  factory Election.fromJson(Map<String, dynamic> json) => Election(
        id: json['id'] as String,
        nameAr: json['name_ar'] as String,
        nameEn: json['name_en'] as String,
        electionType: json['election_type'] as String,
        scheduledDate: json['scheduled_date'] == null ? null : DateTime.parse(json['scheduled_date'] as String),
        isCurrent: json['is_current'] as bool,
        status: json['status'] as String,
      );
}

class ElectionRuleSetSummary {
  final String id;
  final String version;
  final DateTime effectiveFrom;
  final DateTime? effectiveUntil;
  final String electoralSystem;
  final String districtStructure;
  final int totalSeats;
  final double thresholdFraction;
  final String allocationMethod;
  final int minimumCandidateAge;
  final bool allowsIndividualCandidateVotes;
  final String sourceDocument;
  final DateTime verifiedAt;

  const ElectionRuleSetSummary({
    required this.id,
    required this.version,
    required this.effectiveFrom,
    required this.effectiveUntil,
    required this.electoralSystem,
    required this.districtStructure,
    required this.totalSeats,
    required this.thresholdFraction,
    required this.allocationMethod,
    required this.minimumCandidateAge,
    required this.allowsIndividualCandidateVotes,
    required this.sourceDocument,
    required this.verifiedAt,
  });

  /// floor(totalSeats/2) + 1 — never a hard-coded literal in UI code.
  int get majorityThreshold => (totalSeats ~/ 2) + 1;

  factory ElectionRuleSetSummary.fromJson(Map<String, dynamic> json) => ElectionRuleSetSummary(
        id: json['id'] as String,
        version: json['version'] as String,
        effectiveFrom: DateTime.parse(json['effective_from'] as String),
        effectiveUntil: json['effective_until'] == null ? null : DateTime.parse(json['effective_until'] as String),
        electoralSystem: json['electoral_system'] as String,
        districtStructure: json['district_structure'] as String,
        totalSeats: json['total_seats'] as int,
        thresholdFraction: (json['threshold_fraction'] as num).toDouble(),
        allocationMethod: json['allocation_method'] as String,
        minimumCandidateAge: json['minimum_candidate_age'] as int,
        allowsIndividualCandidateVotes: json['allows_individual_candidate_votes'] as bool,
        sourceDocument: json['source_document'] as String,
        verifiedAt: DateTime.parse(json['verified_at'] as String),
      );
}

class TimelineEvent {
  final String id;
  final String milestone;
  final String labelAr;
  final String labelEn;
  final DateTime? startsAt;
  final DateTime? endsAt;

  const TimelineEvent({
    required this.id,
    required this.milestone,
    required this.labelAr,
    required this.labelEn,
    required this.startsAt,
    required this.endsAt,
  });

  factory TimelineEvent.fromJson(Map<String, dynamic> json) => TimelineEvent(
        id: json['id'] as String,
        milestone: json['milestone'] as String,
        labelAr: json['label_ar'] as String,
        labelEn: json['label_en'] as String,
        startsAt: json['starts_at'] == null ? null : DateTime.parse(json['starts_at'] as String),
        endsAt: json['ends_at'] == null ? null : DateTime.parse(json['ends_at'] as String),
      );
}
